pkg/virt-handler/virtwrap/manager.go

```diff
@@ -328,21 +327,14 @@ func VMNamespaceKeyFunc(vm *v1.VM) string {
 	return domName
 }
 
-func (l *LibvirtDomainManager) SyncVM(vm *v1.VM) (*api.DomainSpec, error) {
-	var wantedSpec api.DomainSpec
-	mappingErrs := model.Copy(&wantedSpec, vm.Spec.Domain)
-
-	if len(mappingErrs) > 0 {
-		return nil, errors.NewAggregate(mappingErrs)
-	}
-
+func (l *LibvirtDomainManager) SyncVM(vm *v1.VM, domDesign *designer.DomainDesign) (*api.DomainSpec, error) {
```

Comment:
The device (tap/macvtap/vhostuser) looks rather backend related, than VM OS related. Even if there is finally the need to emulate another VM OS visible device, PCI addresses, MAC addresses, ... can be prefilled. If it is necessary to make another device to the VM OS visible, to allow connecting to a new network, this is acceptable since we can't connect otherwise.

However, something like this is probably not very often the case. My naive assumption is, that in most cases, you still emulate the same interface hardware on the VM OS side (I guess virtio, if the drivers are available, definitely normally PCI. If you want ISA, I  would definitely expect that you as a user added that model to your VM spec, to get that).

So a final flow could look like:
1) user posts a VM spec (maybe with an annotation, saying that it is a linux with virtio support)
2) admission controller A sees that the VM supports virtio, and makes sure, that to all interfaces the model "virtio" is added
3) other admission controller do their work ...
4) Everywhere, where not PCI addresses and mac addresses are specified, the libvirt defaulter would fill in the rest.
5) the VM is accepted by the system and stored in etcd
6) the user gets as a POST response the fully defaulted VM spec (which again does not mean that every possible VM object is filled out)